### PR TITLE
fix(ai): close cloud-AI policy bypass in generateJson/generateImage/streamAiHelpResponse

### PR DESCRIPTION
**Pull request by qnbs**
## Problem

`generateText`/`streamText` in `services/aiProviderService.ts` both apply the cloud-AI privacy/mode policy (`assertCloudAiAllowed`) and local-mode positive routing (`shouldRouteLocally`) before reaching a cloud provider's SDK. Three sibling entry points did not:

- `generateJson`'s `opts.provider === 'gemini'` branch called `generateJsonGemini` directly, with no gate at all.
- `streamAiHelpResponse`'s `opts.provider === 'gemini'` branch had the same gap.
- `generateImage` never called the gate in any branch, including its explicit `'gemini'` case and its `default` case (which silently routed any unrecognized/other provider — e.g. `openrouter` — to Gemini as well).

This is widened by a second gap: the one shared pre-flight policy check in `aiThunkUtils.ts` reads the *global* `advancedAi.provider` setting, while the actual request can resolve to a *different* provider when a per-project AI preset is enabled (`thunkUtils.ts`'s `buildAiOptions` already resolves this correctly). A user with a local global provider and privacy settings requiring local-only AI could have a generation request silently reach Google's real API if the active project had a Gemini-provider preset enabled — the pre-check never inspected the effective provider, and the downstream call sites had no gate of their own either.

## Fix

- Gate all three Gemini-direct branches behind the same `assertCloudAiAllowed` (+ `shouldRouteLocally` where a local fallback path exists) that `generateText`/`streamText` already apply.
- Resolve the thunk pre-check's provider via `buildAiOptions` so it can never diverge from the provider a request actually uses.
- `generateImage`'s `default` case now throws an explicit unsupported-provider error instead of silently calling Gemini — no repository authority (checked `docs/adr/0016-native-grok-and-claude-providers.md` and the image-generation notes in `ROADMAP.md`) defines that behavior as intentional.

## Testing

9 new regression tests across `tests/unit/aiProviderService.test.ts` and `tests/unit/aiThunkUtils.test.ts`. Each guard was mutation-tested individually (temporarily reverted, confirmed the corresponding new test fails, restored). Full existing suite for both files plus `tests/unit/aiProviderService.fallbackChain.test.ts` and `tests/unit/thunks/*` re-run clean, no regressions.

## Non-goals

No catalog changes, no `testAIConnection` changes, no cross-project-mutation fix, no cancellation fix — each is its own separate, causally-independent follow-up.

## Summary by Sourcery

Close cloud-AI policy bypasses and align generation, streaming, routing, and fallback behavior with the effective provider and privacy settings.

Bug Fixes:
- Enforce cloud-AI privacy and local-only policies across Gemini JSON, image, and help-response generation paths, including project-preset and routing-resolved providers.
- Fail explicitly for unsupported image-generation providers instead of silently routing requests to Gemini.
- Improve streaming generation reliability by adding local routing, OpenRouter promotion and fallback handling, cancellation propagation, and duplicate-fallback prevention.

Enhancements:
- Centralize positive routing and OpenRouter fallback resolution for text and streaming generation paths.
- Resolve thunk policy checks against the effective provider used by the request rather than the global provider setting.

Documentation:
- Update the changelog and README test-count references for the added regression coverage.

Tests:
- Add regression coverage for policy enforcement, effective-provider resolution, routing behavior, OpenRouter fallback edge cases, cancellation, and unsupported image providers.

---
## Summary by cubic
Closes a cloud-AI policy bypass where `generateJson`, `generateImage`, and `streamAiHelpResponse` reached Gemini's SDK without the privacy/mode gate that `generateText` and `streamText` already apply. The thunk-level pre-check now validates the effective (preset-aware, routing-resolved) provider, and `streamText` gained the same local/OpenRouter routing parity as `generateText`.

- The shared thunk pre-check resolves the effective provider via side-effect-free `peekPositiveRoutingProvider()` in new `services/ai/positiveRouting.ts`, covering local reroutes and OpenRouter promotion without pulling in provider adapters or logging probe decisions.
- The gemini-direct path in `generateJson`/`streamAiHelpResponse` falls through to the generic path when OpenRouter is preferred, so promotion and fallback handling apply consistently.
- `generateImage` throws an explicit unsupported-provider error instead of silently routing unknown providers to Gemini.
- `streamText` now reroutes local-mode requests to webllm (forwarding the merged `AbortSignal` so cancelled local streams actually stop) and falls back to the configured provider on OpenRouter rate-limit/circuit-open failures, clearing a stale fallback reason when a later request's primary provider succeeds outright.
- OpenRouter fallback in `generateText`/`streamText` swaps the model for webllm, re-checks cancellation, and skips a provider already tried via the promoted fallback.
- Fallback-attempt logic is extracted into shared helpers in new `services/ai/openRouterFallback.ts` so each loop body stays flat and `aiProviderService.ts` does not grow.

**Testing**
- Adds regression tests for each gated branch, preset-aware provider resolution, local rerouting, OpenRouter fallback/promotion, unsupported image providers, and fallback edge cases (including cancellation propagation and stale fallback-reason clearing), plus CHANGELOG and README test-count updates.

<sup>Written for commit a2c249f07296c612f2f46acdae8424fd449b1b53. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/qnbs/WorldScript-Studio/pull/706?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

- **Security**
  - Strengthened cloud-AI privacy safeguards across JSON generation, image generation, and streamed AI help.
  - Local-only settings consistently prevent cloud provider requests while supporting local-mode help.
  - Project AI presets determine the effective provider for policy enforcement.
  - OpenRouter failures use configured fallback providers without duplicate requests.
  - Fallback cancellations are now propagated correctly.
  - Unsupported image-generation providers return explicit errors instead of defaulting to Gemini.

- **Documentation**
  - Updated README test metrics to reflect 7,675+ tests across 605 files.

- **Tests**
  - Expanded coverage for routing, provider selection, fallback behavior, cancellation, and streamed AI help.

___

## **CodeAnt-AI Description**
Enforce AI privacy settings across all generation paths and handle provider fallbacks safely

### What Changed
- Gemini JSON, image, and help-response requests now respect local-only and privacy settings before contacting the cloud service
- Policy checks use the provider that will actually handle the request, including project presets, local routing, and OpenRouter preferences
- Streaming requests now follow local/OpenRouter routing and recover from temporary OpenRouter failures using the configured fallback without duplicate calls
- Image requests for unsupported providers now show an explicit error instead of silently using Gemini
- Added regression coverage and documented the privacy-policy fix

### Impact
`✅ Fewer cloud-AI privacy bypasses`
`✅ Safer local-only generation`
`✅ Fewer streaming failures during OpenRouter outages`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

### CODE REVIEW

File: features/project/aiThunkUtils.ts [hunk 1]
```diff
@@ -66,10 +67,11 @@ export const createDeduplicatedThunk = <Returned, ThunkArg = void>(
...
+        //          of relying on each caller to manually call assertCloudAiAllowed(). Resolves the
+        //          effective (preset-aware) provider the same way buildAiOptions does, so an enabled
+        //          project preset can no longer bypass this check by diverging from the global setting.
         const state = thunkAPI.getState() as RootState;
-        const provider = (state.settings as unknown as { advancedAi?: { provider?: AIProvider } })
-          .advancedAi?.provider;
+        const provider = buildAiOptions(state).provider;
         if (provider) {
           const privacy = (state.settings as unknown as { privacy?: PrivacySettings }).privacy;
           assertCloudAiAllowedSync(provider, privacy);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/aiThunkUtils.ts:75]: Confirmed and fixed: the pre-check now skips assertCloudAiAllowedSync entirely when shouldRouteLocally() is true, since generateText/generateJson already safely reroute a nominally-cloud provider to local inference in that case — rejecting here first would have blocked that path. New regression test proves this.

File: services/aiProviderService.ts [hunk 2]
```diff
@@ -865,7 +870,9 @@ export async function streamAiHelpResponse(
   const helpPromptWithDocs = doc
     ? `You are a helpful assistant for WorldScript Studio. Prefer the documentation excerpts below when they answer the question; otherwise give concise general guidance. Format using Markdown.\n\n${mergedBody}`
     : `You are a helpful assistant for a creative writing app called WorldScript Studio. Answer the user's question concisely and clearly. Format your answer using Markdown. Question: ${sanitizePromptValue(question)}`;
-  if (opts.provider === 'gemini') {
+  // QNBS-v3: gate the Gemini-direct fast path behind the same cloud-AI policy every other entry point applies — this branch previously reached the SDK before any mode/privacy check.
+  if (opts.provider === 'gemini' && !shouldRouteLocally()) {
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/aiProviderService.ts]: Confirmed and fixed: streamText was missing the positive local-routing override that generateText already had. Extracted the shared resolvePositiveRoutingOpts() helper and applied it to both, so streamAiHelpResponse's fallthrough to streamText now correctly reroutes to local inference in local/eco mode instead of rejecting. New regression tests for both streamText and streamAiHelpResponse prove this; the earlier test asserting the reject was updated to assert the correct local-reroute behavior instead.

File: CHANGELOG.md [hunk 3]
```diff
@@ -44,6 +44,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   (`.github/workflows/pr-changelog-reference.yml` + `check-pr-changelog-reference.mjs`, run from the
   PR's base ref to prevent self-weakening) now fails a governed PR's CI unless `[Unreleased]` already
   references it as `PR #<N>`. PR #705.
+- **Closed a cloud-AI privacy-policy bypass reachable via generateJson/generateImage/streamAiHelpResponse:**
+  these three entry points in `aiProviderService.ts` called Gemini directly without the
+  `assertCloudAiAllowed`/`shouldRouteLocally` gate that `generateText`/`streamText` already apply,
+  and `generateImage`'s `default` case silently routed any other provider to Gemini too. Combined
+  with a mismatch where the shared thunk-level policy pre-check read the global provider setting
+  instead of the effective (project-preset-aware) provider, a generation request could silently
+  reach Google's real API despite local-only/privacy settings. All four gaps are now closed. PR #706.
```
[the remark this answers is not in the document]
Comment (qnbs) [reply CHANGELOG.md]: Fixed — reworded the entry to avoid an explicit count that could go stale, and added the two follow-on fixes (pre-check local-routing skip, streamText routing parity) this same finding thread prompted.

File: services/aiProviderService.ts [hunk 4]
```diff
@@ -730,7 +730,9 @@ export async function generateJson<T>(
   signal?: AbortSignal,
 ): Promise<T> {
   try {
-    if (opts.provider === 'gemini') {
+    // QNBS-v3: gate the Gemini-direct fast path behind the same cloud-AI policy every other entry point applies — this branch previously reached the SDK before any mode/privacy check.
+    if (opts.provider === 'gemini' && !shouldRouteLocally()) {
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/aiProviderService.ts]: Same root cause as the codeant-ai finding on this file — fixed by extracting resolvePositiveRoutingOpts() and applying it to streamText. See the reply on that thread for detail.

File: features/project/aiThunkUtils.ts [hunk 5]
```diff
@@ -66,10 +67,11 @@ export const createDeduplicatedThunk = <Returned, ThunkArg = void>(
...
       try {
         // QNBS-v3: Enforce cloud AI policy before every AI thunk — one place to enforce instead
-        //          of relying on each caller to manually call assertCloudAiAllowed().
+        //          of relying on each caller to manually call assertCloudAiAllowed(). Resolves the
+        //          effective (preset-aware) provider the same way buildAiOptions does, so an enabled
+        //          project preset can no longer bypass this check by diverging from the global setting.
         const state = thunkAPI.getState() as RootState;
-        const provider = (state.settings as unknown as { advancedAi?: { provider?: AIProvider } })
-          .advancedAi?.provider;
+        const provider = buildAiOptions(state).provider;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/aiThunkUtils.ts]: Same root cause as the codeant-ai finding on this file — fixed by skipping the pre-check when shouldRouteLocally() is true. See the reply on that thread for detail.

File: features/project/aiThunkUtils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/aiThunkUtils.ts]: Confirmed and fixed — see the replies on the codeant-ai/cubic-dev-ai findings on this same line for the exact fix (skip the pre-check when shouldRouteLocally() is true).

File: services/aiProviderService.ts [hunk 6]
```diff
@@ -865,7 +873,8 @@ export async function streamAiHelpResponse(
   const helpPromptWithDocs = doc
     ? `You are a helpful assistant for WorldScript Studio. Prefer the documentation excerpts below when they answer the question; otherwise give concise general guidance. Format using Markdown.\n\n${mergedBody}`
     : `You are a helpful assistant for a creative writing app called WorldScript Studio. Answer the user's question concisely and clearly. Format your answer using Markdown. Question: ${sanitizePromptValue(question)}`;
-  if (opts.provider === 'gemini') {
+  if (isGeminiDirectCloudPath(opts.provider)) {
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/aiProviderService.ts:912]: Confirmed and fixed — streamText now shares the same resolvePositiveRoutingOpts() positive-routing override generateText already had, so local/eco mode reroutes to webllm instead of rejecting. See the reply on the codeant-ai finding on this file for detail.

File: features/project/aiThunkUtils.ts [hunk 7]
```diff
@@ -65,12 +67,10 @@ export const createDeduplicatedThunk = <Returned, ThunkArg = void>(
...
-        //          of relying on each caller to manually call assertCloudAiAllowed().
+        // QNBS-v3: enforces cloud AI policy for every AI thunk using the effective (preset-aware) provider, skipped when shouldRouteLocally() is true since the real per-call-site gates safely handle that reroute instead.
         const state = thunkAPI.getState() as RootState;
-        const provider = (state.settings as unknown as { advancedAi?: { provider?: AIProvider } })
-          .advancedAi?.provider;
-        if (provider) {
+        const provider = buildAiOptions(state).provider;
+        if (provider && !shouldRouteLocally()) {
           const privacy = (state.settings as unknown as { privacy?: PrivacySettings }).privacy;
           assertCloudAiAllowedSync(provider, privacy);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/aiThunkUtils.ts]: Fixed in b407e825: the thunk policy pre-check now calls resolvePositiveRoutingOpts() (now peekPositiveRoutingProvider(), a side-effect-free variant added in 0a19956f), which applies the same OpenRouter-promotion logic generateText/generateJson/streamText use before checking policy.

File: features/project/aiThunkUtils.ts [hunk 8]
```diff
@@ -2,7 +2,9 @@ import type { AsyncThunkConfig, GetThunkAPI } from '@reduxjs/toolkit';
 import { createAsyncThunk } from '@reduxjs/toolkit';
 import type { RootState } from '../../app/store';
 import { assertCloudAiAllowedSync } from '../../services/ai/aiPolicy';
-import type { AIProvider, PrivacySettings } from '../../types';
+import { resolvePositiveRoutingOpts } from '../../services/aiProviderService';
```
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/aiThunkUtils.ts]: Fixed in 0730e5cb: extracted resolvePositiveRoutingOpts into services/ai/positiveRouting.ts, a new lightweight module that only imports aiModeService.ts and routingLogger.ts (no provider adapters). aiThunkUtils.ts now imports from that module instead of aiProviderService.ts, so the eager startup graph no longer pulls in geminiService/localAiFacade/the OpenRouter provider or the @google/genai chunk.

File: services/aiProviderService.ts [hunk 9]
```diff
@@ -831,11 +853,33 @@ export async function streamText(
...
+            await streamProvider(
+              prompt,
+              creativity,
+              { ...mergedOpts, provider: fallback as AIRequestOptions['provider'] },
+              callbacks,
+              signal,
+            );
+            return;
+          } catch (fallbackError) {
+            lastError = fallbackError;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/aiProviderService.ts:835]: Fixed in 0a19956f: the fallback catch now re-checks isAbortError(fallbackError) (and the merged/outer signal) before storing it as lastError, mirroring the outer catch's existing cancellation guard, and rethrows immediately instead of continuing the chain or reaching a terminal onError.

File: services/aiProviderService.ts [hunk 10]
```diff
@@ -594,16 +594,8 @@ async function generateTextSingleProvider(
...
-  prompt: string,
-  creativity: AiCreativity,
-  opts: AIRequestOptions,
-  signal?: AbortSignal,
-): Promise<string> {
-  // QNBS-v3: Positive routing — apply AI execution mode overrides before dedup keying (G2).
-  // Priority: (1) local-only modes → webllm; (2) OpenRouter enabled → prefer OR for cloud calls;
-  // (3) passthrough — use whatever provider the caller specified.
-  let resolvedOpts = opts;
+// QNBS-v3: positive AI-mode routing (local-only → webllm, else OpenRouter-preferred, else passthrough) shared by generateText and streamText so both reroute before building their fallback chain — streamText was previously missing this step entirely. Also exported for aiThunkUtils.ts's policy pre-check, so it can gate on the same effective provider a real call will actually use.
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/aiProviderService.ts]: Addressed across 0730e5cb (extracting resolvePositiveRoutingOpts out of this file removed several of the flagged comments entirely) and 0a19956f (tightened the remaining thunk-precheck comment to one shorter sentence). The rationale is now spread across fewer, smaller single-purpose comments rather than one large combined one.

File: services/aiProviderService.ts [hunk 11]
```diff
@@ -831,11 +853,33 @@ export async function streamText(
...
+            mode: getActiveAiMode(),
+            originalProvider: 'openrouter',
+            chosenProvider: fallback,
+            reason: 'openrouter-fallback',
+          });
+          try {
+            await streamProvider(
+              prompt,
+              creativity,
+              { ...mergedOpts, provider: fallback as AIRequestOptions['provider'] },
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/aiProviderService.ts]: Fixed in 0a19956f: extracted buildOpenRouterFallbackOpts(), which swaps in getLocalFallbackModel() when the fallback provider is webllm (Gemini's own getModelForText() already sanitizes an unrecognized model id, so only webllm needed an explicit swap). Applied to both generateText and streamText.

File: services/aiProviderService.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (qnbs) [reply services/aiProviderService.ts:835]: Fixed in 0a19956f: same fix as the sibling cancellation finding on this file (chatgpt-codex-connector) — the fallback catch now rethrows immediately on isAbortError/aborted signal instead of storing it as lastError.

File: services/aiProviderService.ts [hunk 12]
```diff
@@ -831,11 +811,33 @@ export async function streamText(
...
+              callbacks,
+              signal,
+            );
+            return;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/aiProviderService.ts:827]: Fixed in 0a19956f: streamText's OpenRouter-fallback success branch now sets _lastFallbackReason before returning, mirroring generateText's existing behavior.

File: services/aiProviderService.ts [hunk 13]
```diff
@@ -831,11 +811,33 @@ export async function streamText(
...
+          } catch (fallbackError) {
+            lastError = fallbackError;
+          }
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/aiProviderService.ts:836]: Fixed in 0a19956f: both generateText and streamText now track the OpenRouter-promoted fallback provider (attemptedOpenRouterFallback) and skip a matching later chain entry, so a provider already tried via promotion is never invoked a second time from the outer loop.

File: features/project/aiThunkUtils.ts [hunk 14]
```diff
@@ -65,11 +67,9 @@ export const createDeduplicatedThunk = <Returned, ThunkArg = void>(
       } as DeduplicatedThunkAPI;
 
       try {
-        // QNBS-v3: Enforce cloud AI policy before every AI thunk — one place to enforce instead
-        //          of relying on each caller to manually call assertCloudAiAllowed().
+        // QNBS-v3: enforces cloud AI policy using resolvePositiveRoutingOpts()'s effective provider — the same function generateText/generateJson/streamText use to actually route the request — so a local- or OpenRouter-promoted call is checked against what it will really use, not the raw preset/global setting.
         const state = thunkAPI.getState() as RootState;
-        const provider = (state.settings as unknown as { advancedAi?: { provider?: AIProvider } })
-          .advancedAi?.provider;
+        const provider = resolvePositiveRoutingOpts(buildAiOptions(state)).provider;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/aiThunkUtils.ts]: Fixed in 0a19956f: split resolvePositiveRoutingOpts into a pure computeRoutingResolution() plus two thin wrappers — resolvePositiveRoutingOpts() (logs, used by real generation calls) and the new peekPositiveRoutingProvider() (no logging, used by the thunk pre-check). The pre-check no longer records a routing-decision entry for a probe that may not reflect what actually executes.

File: services/ai/positiveRouting.ts [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    logRoutingDecision({
+      mode: getActiveAiMode(),
+      originalProvider: opts.provider,
+      chosenProvider: 'webllm',
+      reason: 'mode-override',
+    });
+    return { ...opts, provider: 'webllm', model: localModel as AIRequestOptions['model'] };
+  }
+  if (
+    shouldUseOpenRouter() &&
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/ai/positiveRouting.ts]: Fixed in 0a19956f: isGeminiDirectCloudPath() now also excludes shouldUseOpenRouter(), so generateJson (and streamAiHelpResponse) fall through to the generic generateText-based path -- which does apply resolvePositiveRoutingOpts's OpenRouter promotion -- instead of taking the direct Gemini SDK shortcut when OpenRouter is preferred.

File: features/project/aiThunkUtils.ts [hunk 16]
```diff
@@ -65,11 +67,9 @@ export const createDeduplicatedThunk = <Returned, ThunkArg = void>(
...
         if (provider) {
```
[the remark this answers is not in the document]
Comment (qnbs) [reply features/project/aiThunkUtils.ts]: Reviewed against current code, not implementing a change here: generateImage's real gate (generateImageViaGemini -> assertCloudAiAllowed('gemini')) fires unconditionally for the actual provider it is about to use, regardless of what the thunk pre-check predicted -- so a local-only-mode image request that the pre-check (correctly) approved as 'webllm' still gets rejected at generateImage's own boundary before any network call, and an EU-residency-restricted request behaves the same way. There's no reachable path where a disallowed image request slips through; the mismatch is a pre-check prediction accuracy/UX nit, not a policy bypass. Making the pre-check image-aware would need a per-thunk feature/capability tag createDeduplicatedThunk doesn't have today (image is the only feature with no local-routing fallback) -- a larger, out-of-scope change for this PR. Leaving this open for a possible follow-up if the maintainer wants that capability-tagging built.

File: services/aiProviderService.ts [hunk 17]
```diff
@@ -831,11 +870,36 @@ export async function streamText(
...
+          try {
+            await attemptOpenRouterStreamFallback(
+              prompt,
+              creativity,
+              mergedOpts,
+              callbacks,
+              fallback,
+              signal,
+            );
+            _lastFallbackReason = `OpenRouter rate-limited; fell back to ${fallback}.`;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/aiProviderService.ts:826]: Fixed in a2c249f0: streamText's per-provider success path now mirrors generateText's exact _lastFallbackReason bookkeeping (clears it on an immediate primary success, records a reason when a later chain entry succeeds), so a stale reason from an earlier request no longer survives a subsequent, fully successful primary-provider stream.

File: services/aiProviderService.ts [hunk 18]
```diff
@@ -794,10 +826,17 @@ export async function streamText(
   callbacks: AIStreamCallbacks,
   signal?: AbortSignal,
 ): Promise<void> {
-  const { key, controller } = _deduplicateRequest(opts.provider, opts.model, prompt);
-  const mergedOpts = withMergedAbortSignal(opts, signal ?? controller.signal);
+  const resolvedOpts = resolvePositiveRoutingOpts(opts);
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/aiProviderService.ts:768]: Fixed in a2c249f0: the webllm/onnx/transformers branch in streamProvider now forwards oWithLora.signal into generateLocalText's 5th parameter, so a cancelled or offline-rerouted local stream actually stops instead of running to completion.

File: services/ai/positiveRouting.ts [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    originalProvider: opts.provider,
+    chosenProvider: resolvedOpts.provider,
+    reason,
+  });
+  return resolvedOpts;
+}
+
+// QNBS-v3: side-effect-free variant for the thunk policy pre-check — the real call re-resolves (and logs) again when it actually executes, so this must not double-log or record a decision for a request that may never be dispatched with this exact shape (e.g. generateImage, which never calls resolvePositiveRoutingOpts itself).
+export function peekPositiveRoutingProvider(opts: AIRequestOptions): AIRequestOptions['provider'] {
+  return computeRoutingResolution(opts).resolvedOpts.provider;
```
[the remark this answers is not in the document]
Comment (qnbs) [reply services/ai/positiveRouting.ts:73]: Reviewed against current code, no change made: generateImage() dispatches purely on the raw opts.provider it receives (never consults resolvePositiveRoutingOpts/the thunk pre-check's resolved value), and its only real path -- generateImageViaGemini -- unconditionally calls assertCloudAiAllowed('gemini') before any network call. So even when the pre-check approves a hybrid-mode/shouldRouteLocally() request as 'webllm', generateImage itself never substitutes webllm for image generation and never skips its own policy gate against the actual provider (gemini) it uses -- the capability invariant (no silent Gemini fallback, no pretending a local runtime supports image generation) already holds. The remaining gap is pre-check *prediction* accuracy for this one non-text feature, not a capability or policy bypass; closing that gap fully would need a per-thunk feature/capability tag that createDeduplicatedThunk doesn't have today, which is out of scope for this PR. Tracking as a follow-up, not a #706 blocker.
